src/kernels/MIOpenReduceLogicalCalculation.cpp

```diff
@@ -0,0 +1,141 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "miopen_cstdint.hpp"
+#include "MIOpenReduceCalculation.hpp"
+
+template <typename TI, ReduceCalculationOp_t op>
+__device__ void logical_calculationparallelfwdcontiguous(const TI* __restrict__ x,
+                                                         bool* __restrict__ y,
+                                                         const uint64_t output_numel,
+                                                         const uint64_t reduce_size,
+                                                         const uint64_t parallelism_size,
+                                                         const uint64_t inner_size)
+{
+    const uint64_t gid = threadIdx.x + blockIdx.x * blockDim.x;
+    if(gid >= parallelism_size * output_numel)
+        return;
+
+    uint64_t n = inner_size * parallelism_size;
+
+    uint64_t slice_id       = gid / n;
+    uint64_t slice_local_id = gid % n;
+
+    uint64_t input_idx = slice_id * inner_size * reduce_size + slice_local_id;
+
+    uint64_t parallel_id = slice_local_id / inner_size;
+#if MIOPEN_USE_INT8
+    TI calculation = reduce_func<TI, op>{}.get_initial_value();
+    for(uint64_t k = parallel_id; k < reduce_size; k += parallelism_size)
+    {
+        TI val = x[input_idx];
+        reduce_func<TI, op>{}.calculate(calculation, val);
+        input_idx += inner_size * parallelism_size;
+    }
+
+    y[gid] = calculation == 0 ? 0 : 1;
+#else
+    FLOAT_ACCUM calculation = reduce_func<FLOAT_ACCUM, op>{}.get_initial_value();
+    for(uint64_t k = parallel_id; k < reduce_size; k += parallelism_size)
+    {
+        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(x[input_idx]);
+        if(isnan(val))
+        {
+            val = static_cast<FLOAT_ACCUM>(1);
+        }
+        reduce_func<FLOAT_ACCUM, op>{}.calculate(calculation, val);
+        input_idx += inner_size * parallelism_size;
+    }
+
+    y[gid]                  = CVT_ACCUM2FLOAT(calculation) == 0 ? 0 : 1;
```

Comment:
use template type to handle this instead of global variable `MIOPEN_USE_INT` to avoid potential error

Reply:
solved